File: mapox/wrappers/vocab_wrapper.py
from typing import Any
from functools import cached_property

import jax
from jax import numpy as jnp

from mapox.environment import Environment, EnvState
from mapox.envs.common import make_obs_spec
from mapox.renderer import GridRenderState, GridRenderSettings
from mapox.specs import ActionSpec, DiscreteActionSpec, ObservationSpec
from mapox.timestep import TimeStep
from mapox.vocab import Vocabulary


class VocabWrapper(Environment):
    """Presents a local-vocab env in a global vocabulary.

    Global action ids in, global obs/action ids out; translation happens at
    the boundary. Global symbols the inner env lacks map to local id 0 (they
    are masked out, so the agent never legally selects them).
    """

    def __init__(
        self,
        env: Environment,
        action_vocab: Vocabulary,
        obs_vocab: Vocabulary,
    ):
        self._env = env
        self._action_vocab = action_vocab
        self._obs_vocab = obs_vocab

        self._global_to_local_action = action_vocab.lut_to(
            env.action_vocab, default=0
        )
        self._local_to_global_action = env.action_vocab.lut_to(
            action_vocab, default=0
        )
        # int8 so tile-channel scatter matches the obs dtype; make_obs_spec
        # guarantees the vocab fits.
        self._local_to_global_obs = env.obs_vocab.lut_to(
            obs_vocab, default=0, dtype=jnp.int8
        )

    def __getattr__(self, name):
        return getattr(self._env, name)

    def _encode_timestep(self, ts: TimeStep) -> TimeStep:
        mask = jnp.zeros(
            (*ts.action_mask.shape[:-1], len(self._action_vocab)), jnp.bool_
        )
        mask = mask.at[..., self._local_to_global_action].set(ts.action_mask)

        # only the first obs channel goes through the lookup
        obs = ts.obs.at[..., 0].set(self._local_to_global_obs[ts.obs[..., 0]])

        return ts._replace(
            last_action=self._local_to_global_action[ts.last_action],
            obs=obs,
            action_mask=mask,
        )

    def reset(self, rng_key: jax.Array):
        state, ts = self._env.reset(rng_key)
        return state, self._encode_timestep(ts)

    def step(self, state, action: jax.Array, rng_key: jax.Array):
        state, ts = self._env.step(
            state, self._global_to_local_action[action], rng_key
        )
        return state, self._encode_timestep(ts)

    @cached_property
    def observation_spec(self) -> ObservationSpec:
        spec = self._env.observation_spec
        if spec.max_value is None:
            raise ValueError("Obs must have a max value")
        if len(spec.shape) != 3 or spec.shape[2] != 4:
            raise ValueError("VocabWrapper only supports obs of (width, height, 4)")

        return make_obs_spec(spec.shape[0], spec.shape[1], len(self._obs_vocab))

    @cached_property
    def action_spec(self) -> ActionSpec:
        return DiscreteActionSpec(len(self._action_vocab))

    @property
    def obs_vocab(self) -> Vocabulary:
        return self._obs_vocab

    @property
    def action_vocab(self) -> Vocabulary:
        return self._action_vocab

    @property
    def num_agents(self) -> int:
        return self._env.num_agents

    @property
    def teams(self) -> jax.Array | None:
        # Environment.teams has a concrete default, so __getattr__ never
        # delegates it; forward explicitly.
        return self._env.teams

    def create_placeholder_logs(self) -> dict[str, Any]:
        return self._env.create_placeholder_logs()

    def create_logs(self, state) -> dict[str, Any]:
        return self._env.create_logs(state)

    def get_render_settings(self) -> GridRenderSettings:
        return self._env.get_render_settings()

    def get_render_state(self, state: EnvState) -> GridRenderState:
        return self._env.get_render_state(state)

File: crates/mapox-core/src/wrappers/vocab_wrapper.rs
use ndarray::Array2;

use crate::{
    env::Environment,
    render::env::{GridRenderSettings, GridRenderState},
    spec::{ActionSpec, ObservationSpec},
    timestep::TimeStepMut,
    vocab::{VocabId, Vocabulary},
};

pub struct VocabWrapper {
    action_vocab: Vocabulary,
    observation_vocab: Vocabulary,
    global_to_local_action: Vec<VocabId>,
    local_to_global_action: Vec<VocabId>,
    local_to_global_obs: Vec<VocabId>,
    inner: Box<dyn Environment>,
    temp_actions: Vec<VocabId>,
    temp_action_mask: Array2<bool>,
}

impl VocabWrapper {
    pub fn new(
        inner: Box<dyn Environment>,
        action_vocab: &Vocabulary,
        observation_vocab: &Vocabulary,
    ) -> Self {
        let num_global_actions = action_vocab.len();
        let num_agents = inner.num_agents();

        Self {
            global_to_local_action: action_vocab.lut_to(inner.action_vocab(), 0),
            local_to_global_action: inner.action_vocab().lut_to(action_vocab, 0),
            local_to_global_obs: inner.obs_vocab().lut_to(observation_vocab, 0),
            action_vocab: action_vocab.clone(),
            observation_vocab: observation_vocab.clone(),
            inner,
            temp_actions: vec![0; num_agents],
            temp_action_mask: Array2::from_elem((num_agents, num_global_actions), false),
        }
    }

    fn encode_timestep(&mut self, timestep: &mut TimeStepMut) {
        self.temp_action_mask.assign(&timestep.action_mask);
        timestep.action_mask.fill(false);

        for agent_id in 0..self.num_agents() {
            for local_action in 0..self.inner.action_vocab().len() {
                let mask = self.temp_action_mask[[agent_id, local_action]];
                let global_action = self.local_to_global_action[local_action];
                timestep.action_mask[[agent_id, global_action as usize]] = mask;
            }
        }

        timestep
            .last_action
            .map_inplace(|action| *action = self.local_to_global_action[*action as usize]);

        timestep
            .obs
            .map_inplace(|obs| *obs = self.local_to_global_obs[*obs as usize]);
    }
}

impl Environment for VocabWrapper {
    fn reset(&mut self, seed: u64, timestep: &mut TimeStepMut) {
        self.inner.reset(seed, timestep);
        self.encode_timestep(timestep);
    }

    fn step(&mut self, actions: &[VocabId], timestep: &mut TimeStepMut) {
        for (&local, target) in actions.iter().zip(self.temp_actions.iter_mut()) {
            *target = self.global_to_local_action[local as usize];
        }

        self.inner.step(&mut self.temp_actions, timestep);
        self.encode_timestep(timestep);
    }

    fn action_spec(&self) -> ActionSpec {
        ActionSpec {
            num_actions: self.action_vocab.len(),
        }
    }

    fn observation_spec(&self) -> ObservationSpec {
        ObservationSpec {
            num_types: self.observation_vocab.len(),
            ..self.inner.observation_spec()
        }
    }

    fn action_vocab(&self) -> &Vocabulary {
        &self.action_vocab
    }

    fn obs_vocab(&self) -> &Vocabulary {
        &self.observation_vocab
    }

    fn num_agents(&self) -> usize {
        self.inner.num_agents()
    }

    fn get_render_settings(&self) -> GridRenderSettings {
        self.inner.get_render_settings()
    }

    fn render_state_into(&self, grid_render_state: &mut GridRenderState) {
        self.inner.render_state_into(grid_render_state)
    }

    fn num_tasks(&self) -> usize {
        self.inner.num_tasks()
    }
}
